server/plugin.go

```diff
@@ -0,0 +1,147 @@
+package main
+
+import (
+	"io"
+	"net/http"
+	"time"
+
+	"github.com/gorilla/mux"
+	"github.com/ilgooz/mattermost-plugin-topdf/server/gotenberg"
+	"github.com/ilgooz/mattermost-plugin-topdf/server/topdf"
+	"github.com/ilgooz/mattermost-plugin-topdf/server/x/xhttp"
+	"github.com/ilgooz/mattermost-plugin-topdf/server/x/xtime"
+	"github.com/mattermost/mattermost-server/plugin"
+	"github.com/rs/cors"
+)
+
+// Plugin is a TOPDF app that implements plugin.MattermostPlugin.
+type Plugin struct {
+	plugin.MattermostPlugin
+
+	// app is the actual, underlying TOPDF app and its features exposed to
+	// network via Plugin's HTTP API.
+	app interface {
+		CheckServerStatus() (running bool, err error)
+		GetPDF(userID, fileID string) (pdf io.ReadCloser, err error)
+	} // *topdf.TOPDF
+
+	// gt used by the app to interact with topdf.PDFServer implementation(Gotenberg).
+	// it's also used by Plugin to update Gotenberg's configs when Plugin configs reloaded.
+	gt *gotenberg.Gotenberg
```

Comment:
Given the (nice) abstraction between the plugin and the app, it seems unfortunate to have to expose the Gotenberg layer at all here. Should we instead surface the necessary API on the `TOPDF` struct instead?

Reply:
I was liking this approach because it's easier to mock Gotenberg's funcs this way while testing `topdf` pkg without the need of faking in a lower, `httptest` layer.

--

Otherwise, if we still want to mock Gotenberg's funcs, we need to create multiple interfaces around.

For example, `gotenberg.New()` has to return an interface instead of `*gotenberg.Gotenberg` and `topdf.New()` has to accept an interface with `gotenberg.New()` func in it.

I kind of don't like taking this approach since it feels a bit not idiomatic to me.

--

In the `main` plugin package, I think of `*gotenberg.Gotenberg` like a dependency to `*topdf.TOPDF` and it feels very natural to me.

Also this way, it is very easy to configure dependencies without the need to proxying configurations to the dependency via it's dependent, in this case `topdf.TOPDF`.

I remember that I first saw this technique in go-micro a long time ago and adopted since.

What do you think how we should proceed on this one?

--

P.S. from [go-micro](https://micro.mu/docs/framework.html#plugin-option):
```go
package main

import (
        "github.com/micro/go-micro" 
        // consul registry
        "github.com/micro/go-plugins/registry/consul"
        // rabbitmq transport
        "github.com/micro/go-plugins/transport/rabbitmq"
        // kafka broker
        "github.com/micro/go-plugins/broker/kafka"
)

func main() {
	registry := consul.NewRegistry()
	broker := kafka.NewBroker()
	transport := rabbimq.NewTransport()

        service := micro.NewService(
                micro.Name("greeter"),
                micro.Registry(registry),
                micro.Broker(broker),
                micro.Transport(transport),
        )

	service.Init()
	service.Run()
}
```